File: src/advent2020/day4/part1.js
//const input = await Deno.readTextFile('./input_light.txt');
const input = await Deno.readTextFile('./input.txt');

console.log('test');

const plines = input.split('\n\n');
console.log(plines);

function getPassport(line) {
  const p = line
    .split('\n')
    .join(' ')
    .split(' ')
    .map((l) => l.split(':'))
    .reduce((acc, l) => {
      if (!l[0]) return acc;
      acc[l[0]] = l[1];
      return acc;
    }, {});
  console.log(p);
  return p;
}

function checkFields(p) {
  // byr (Birth Year)
  // iyr (Issue Year)
  // eyr (Expiration Year)
  // hgt (Height)
  // hcl (Hair Color)
  // ecl (Eye Color)
  // pid (Passport ID)
  // cid (Country ID)
  const fields = ['byr', 'iyr', 'eyr', 'hgt', 'hcl', 'ecl', 'pid'];
  const res = fields.map((f) => p[f]).filter(Boolean);
  console.log({ p, res });
  return res.length === fields.length;
}
console.log(plines.map(getPassport).map(checkFields).filter(Boolean).length);

File: src/advent2020/day4/part1.clj
(ns advent2020.day4.part1
  (:require [clojure.core :refer [slurp]]
            [clojure.java.io :as io]
            [clojure.string :as str]))

(def data (-> "day4/input_light.txt"
              io/resource
              slurp
              (str/split #"\n\n")))

data

;;(str/split
(defn str->pairs [item]
  (->> (str/split item #"[ \n]")
       (map #(str/split % #":"))
       (map (fn [ [ a b ] ] [ (keyword "day04" a) b]))
       ))

(def data-map (->> (map str->pairs data)
                   (map #( into {} %))
                   (map #(dissoc % :day04/cid))
                   ))
(str->pairs (first data))
data-map

(def field-strings ["byr" "yir" "eyr" "hgt" "hcl" "ecl" "pid"])
(def fields (set (map #(keyword "day04" %) field-strings)))

(fields (first data-map))
data-map
(defn part1 [d]
  (->> d
       (filter #(= (set (keys %)) fields))
;;       count
       ))

(part1 data-map)
